### ISSUE TITLE
[WebGPU EP] GridSample kernel produces incorrect output in deformable-attention decoders (RT-DETR-family)

### ISSUE THREAD

**Issue by sjvrensburg** [microsoft/onnxruntime#32275]
> **Note:** This report was drafted with AI assistance (Claude Code, Anthropic) while investigating a
> downstream bug in [RailReaderCore](https://github.com/sjvrensburg/RailReaderCore). The
> investigation, data, and conclusions below were reviewed by the human reporter (repo owner)
> before submission. Flagging this openly rather than hiding it.

### Describe the issue

The WebGPU execution provider's `GridSample` kernel appears to produce numerically incorrect
results relative to the CPU EP, specifically inside deformable-attention decoders (RT-DETR /
Deformable-DETR-family object detection models). Everything upstream of the first `GridSample`
node in the graph — a ResNet-style CNN backbone, a transformer encoder layer, several
LayerNormalization ops — matches the CPU EP output at cosine similarity 0.9999–1.0. The moment
the graph reaches the first `GridSample` op, cosine similarity between the WebGPU EP and CPU EP
outputs collapses to ~0.52 and does not recover through the rest of the decoder.

This was surfaced as a real-world symptom first: an RT-DETRv2 layout-detection model
([Docling Heron](https://huggingface.co/docling-project/docling-layout-heron-onnx), FP16 export)
run via the WebGPU EP silently produces far fewer/lower-quality detections than the same model
on the CPU EP — roughly half the boxes, with the majority of the discrepancy being boxes that
*are* detected on GPU but shifted/shrunk enough to fall below a 0.5 IoU match threshold, not
outright absent. A second, architecturally unrelated model
([PaddlePaddle PP-DocLayoutV3](https://huggingface.co/PaddlePaddle/PP-DocLayoutV3_safetensors),
also FP16, also RT-DETR-family) shows the same class of degradation. Both models use
`GridSample` for deformable attention (Heron: 18 nodes, 6 decoder layers × 3; V3: also 18 nodes).

Isolating the execution provider as the sole variable (running the *identical* FP16 weights on
CPU EP vs WebGPU EP, so FP16 storage precision is controlled for) reproduces essentially the same
magnitude of divergence as comparing against a full-precision CPU baseline — ruling out FP16
precision loss as the cause and pointing at the WebGPU EP's own kernel computation.

Layer-by-layer activation diff on Heron-FP16 (representative page, `14-ar-ma-models.pdf` p1):

```
output                                                          n        meanAbs    maxAbs    cosSim
stem Conv                                                  3276800   7.2e-04   1.6e-02   1.00000
backbone ~25% Conv                                          819200   1.7e-03   2.3e-02   0.99992
backbone ~50% Conv                                          819200   6.1e-04   1.3e-02   0.99997
decoder_input_proj (end of backbone)                        102400   8.8e-04   3.5e-02   0.99999
encoder.0 self_attn LayerNorm (encoder start)                102400   3.4e-03   3.7e-02   0.99999
decoder/layers.0/encoder_attn/GridSample                    307200   3.2e-01   3.1e+00   0.51835  <-- divergence starts here
decoder/layers.5/encoder_attn/GridSample_2 (last layer)      307200   1.5e-01   1.3e+00   0.52872
decoder/layers.5/final_layer_norm                            76800   2.6e-01   8.6e+00   0.84172
```

The step-function nature of the divergence (near-perfect agreement, then a sharp drop exactly at
the first `GridSample` node, with no gradual precision drift beforehand) points at a correctness
bug in the kernel rather than accumulated FP16 rounding.

Possibly relevant: the WebGPU EP's `GridSample` kernel was only added in #28264 (merged
2026-04-29), a fairly young addition; the PR author's own description notes limited testing
("I haven't tested that the onnxruntime-web build would pick up this new operator
implementation"). I could not find an existing open issue describing a WebGPU `GridSample`
correctness regression.

### To reproduce

Full reproduction tooling (C#, .NET 10) is here:
https://github.com/sjvrensburg/RailReaderCore/tree/feat/gpu-threshold-calibration-109/tools/webgpu-diag

- `make_debug_model.py` adds extra graph outputs at the checkpoints shown above to a source
  ONNX file (uses only the `onnx` Python package — no onnxruntime needed for this step).
- `Program.cs` loads the resulting debug model twice — once with the default CPU EP, once with
  the WebGPU EP (`OrtEnv.RegisterExecutionProviderLibrary` + `SessionOptions.AppendExecutionProvider`,
  via `Microsoft.ML.OnnxRuntime.EP.WebGpu` 0.3.0) — runs both on byte-identical preprocessed
  input, and reports mean/max absolute difference and cosine similarity per output tensor.

Minimal repro shape, if a smaller isolated `GridSample`-only test case would help narrow this
further, I'm happy to build one — the RT-DETR decoder's `GridSample` calls use `align_corners:
false`, `mode: "bilinear"`, `padding_mode: "zeros"`, on a 4D grid; I can extract the exact
op attributes/shapes from the graph if useful.

Any full model file needed for a completely self-contained repro can be provided —
the FP16 Heron export used above is at
https://huggingface.co/stefanj0/docling-layout-heron-fp16-onnx (public).

### Urgency

Not urgent/blocking a release deadline, but it silently produces materially wrong output (not a
crash), which is worse than a hard failure for anyone relying on WebGPU EP correctness for
detection/segmentation models using deformable attention (RT-DETR, Deformable-DETR, and any
architecture using `GridSample`).

### System information

- **Platform:** Linux (Ubuntu 24.04.4 LTS, kernel 7.0.0-30-generic)
- **ONNX Runtime Installation:** Released Package (NuGet)
- **ONNX Runtime Version:** `Microsoft.ML.OnnxRuntime` 1.24.4 + `Microsoft.ML.OnnxRuntime.EP.WebGpu` 0.3.0
- **ONNX Runtime API:** C#
- **Architecture:** X64
- **Execution Provider:** WebGPU (native plugin EP via `Microsoft.ML.OnnxRuntime.EP.WebGpu`; Dawn → Vulkan on Linux — not in the EP dropdown list above, hence "Other")
- **Execution Provider Library Version:** Vulkan instance 1.3.275, NVIDIA RTX A2000 8GB Laptop GPU, driver 580.173.02 (Intel Iris Xe iGPU also present on the same machine, not confirmed which device WebGPU picked)

**Comment by sjvrensburg** [comment]
### Update: not a GridSample kernel bug — closing

I dug further into this (using the actual `PP-DocLayoutV3-fp16.onnx` weights, running both CPU EP and the real native WebGPU plugin EP side-by-side with debug outputs added at every node on the path into and out of `GridSample`), and the root cause is different from what I originally reported. Posting the corrected diagnosis for anyone who finds this issue later.

**The `GridSample` WGSL kernel itself is correct.** I extracted the exact shader `onnxruntime/core/providers/webgpu/tensor/grid_sample.cc` generates for `mode=bilinear, padding_mode=zeros, align_corners=false` and ran it directly on real Vulkan hardware (via `wgpu-py`) against a NumPy bilinear-sampling reference, at RT-DETR-realistic shapes (C=256, H_out=300, W_out=4), in both fp32 and fp16. It matched to floating-point precision in every case — no discrepancy.

**Bisecting the real model pinpointed the actual source:** `GridSample`'s own *inputs* already diverge from the CPU EP before the op even runs (the sampling-grid input was at cosine similarity ~0.49 going in). Tracing that back through the graph:

- The encoder's `TopK` op (selecting the top-300 query proposals by objectness score) is where the real divergence starts.
- The scores feeding `TopK` match CPU vs GPU almost exactly (cosine similarity 1.00000, differences at the fp16-rounding level, mean abs ≈ 0.0078).
- But those tiny rounding differences are enough to flip the tie-break order right at the top-300 cutoff: 295/300 selected indices matched between CPU and GPU, and the mismatches were adjacent-rank swaps of near-tied scores (e.g. CPU picks `[…,12784,12785,…]`, GPU picks `[…,12785,12784,…]`).
- Every downstream per-query computation (mask head, box decode via a `Greater`-than-zero threshold + `ReduceMin`/`ReduceMax` over mask logits, then `GridSample` in the decoder's deformable attention) is keyed to query *position*. A handful of reordered/swapped queries at the selection boundary, combined with the `Greater` threshold's sensitivity to sub-fp16-epsilon score differences near mask boundaries, cascades into large elementwise tensor differences downstream — even though the underlying feature maps were correct to 0.9999+ all the way through the backbone and encoder. `GridSample` just happened to be the first checkpoint I'd instrumented downstream of that cascade, which is why it looked like the divergence "started" there.

So: this is ordinary fp16 cross-backend rounding noise, amplified by a hard `TopK` selection boundary and a `Greater`-than-zero mask threshold — not a correctness bug in any single kernel, and not something fixable inside `grid_sample.cc`. Apologies for the noise — closing this out. Happy to share the debug-model / bisection harness if it's useful for anyone hitting similar CPU-vs-GPU divergence on TopK-based detection models.
